Telegram bot as a platform (both webhook and polling)

sander76 opened the pull request:
## Description:
Telegram bot as a platform with the option to choose for a polling or a webhook method.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#<home-assistant.github.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml

telegram_bot:
    platform: telegram_webhooks or telegram_poll
    api_key : api_key_here
    allowed_chat_ids : 
        - 123456
        - 456789


```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.


[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

sander76 committed:
rebasing from dev.

sander76 committed:
rebasing from dev.

sander76 committed:
first commit

sander76 committed:
first commit

sander76 committed:
going over the checklist with things to do..

sander76 committed:
going over the checklist with things to do..

sander76 committed:
going over the checklist with things to do..

sander76 committed:
going over the checklist with things to do..

sander76 committed:
removing forgotten big comment block.

sander76 committed:
removing forgotten big comment block.

sander76 committed:
removing forgotten big comment block.

sander76 commented:
@scipioni I hope this to be the proper implementation of telegram-webhooks and telegram-poll as a platform. Are you able to give the webhooks part a try ?

sander76 commented:
@balloob No one seems to be assigned as reviewer. Is that correct ? I am wondering whether I am on the right track with my PR....

balloob commented on homeassistant/components/telegram_bot/telegram_poll.py:
```diff
@@ -0,0 +1,120 @@
+        if offset:
+            data['offset'] = offset
+        try:
+            resp = yield from self.websession.post(
```
You want to wrap this in a timeout, or else this request to the server might never end. See an example here: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/camera/generic.py#L113

sander76 replied:
Thought about that, but I figured the timeout in the request body would take care of that. Adding an extra one looked just strange. But I can add one with a timeout slightly longer than the one in the request body...

balloob commented on homeassistant/components/telegram_bot/telegram_poll.py:
```diff
@@ -0,0 +1,120 @@
+        _updates = yield from self.get_updates(self.update_id, 10)
+        for update in _updates['result']:
+            self.update_id = update['update_id'] + 1
+            event, event_data = yield from process_message(
```
I would `process_message` to also fire the message over the event bus. No need to duplicate that logic among the two platforms?

balloob commented on homeassistant/components/telegram_bot/telegram_poll.py:
```diff
@@ -0,0 +1,120 @@
+                # timeout will for this reason not really stress the processor.
+                yield from self.handle()
+        except CancelledError:
+            _LOGGER.error("Stopping telegram polling")
```
This should not be logged as `error`.

balloob commented on homeassistant/components/telegram_bot/__init__.py:
```diff
@@ -0,0 +1,82 @@
+@asyncio.coroutine
+def async_setup(hass, config):
+    """Track states and offer events for sensors."""
+    component = EntityComponent(_LOGGER, DOMAIN, hass)
```
Please do not use the EntityComponent but use the few building blocks that the EntityComponent uses that you need. See the notify component for an example: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/notify/__init__.py#L86-L112

balloob reviewed (changes requested):
There is no reviewer assigned until someone has time to review it and does so. And that all depends on when people feel like it.

Left some comments

sander76 commented:
@balloob. Thanks man! Normally I get the balloobbot (if I remember correctly) mentioning people who are probably best fit for reviewing. Not seeing it here I thought something might be wrong...

balloob commented:
Ah yes, that is @mention-bot. Not sure why it didn't run.

andrey-git commented:
@sander76 Do you plan to continue working on this PR?

sander76 commented:
@andrey-git I certainly am ! I guess I am through 90% of the changes balloob suggested...

sander76 commented:
@andrey-git If you're able to help me check whether the webhook platform works that would be great.....

andrey-git commented:
I don't have telegram, sorry :)

sander76 commented:
@andrey-git Ah, so the question has more to do with cleaning up stale pull requests ?

andrey-git commented:
Yes, the PR was quiet for 2 weeks, so I wanted to check if it was still relevant.

sander76 committed:
changing setup methods.
wrapping post method in async timeout
and other changes suggested by balloob.

sander76 committed:
changing setup methods.
wrapping post method in async timeout
and other changes suggested by balloob.

sander76 committed:
changing setup methods.
wrapping post method in async timeout
and other changes suggested by balloob.

sander76 committed:
changing setup methods.
wrapping post method in async timeout
and other changes suggested by balloob.

sander76 committed:
changing setup methods.
wrapping post method in async timeout
and other changes suggested by balloob.
trying to fix .coveragerc

sander76 commented:
Closing. Feeling frustrated. Github is annoying me to hell here.
